test/shanghai-opcodes/ShanghaiOpcodes.js

```diff
@@ -0,0 +1,51 @@
+const { expect } = require("chai");
+const { ethers } = require("hardhat");
+
+describe("Shanghai opcodes tests", function () {
+  let signers;
+  let shanghaiContract;
+
+  before(async function () {
+    signers = await ethers.getSigners();
+
+    const factory = await ethers.getContractFactory('ShanghaiOpcodes');
+    shanghaiContract = await factory.deploy();
+  });
+
+  it("should be able to execute opShl()", async function () {
+    const res = await shanghaiContract.opShl(2, 10);
+    // shift left
+    expect(res).to.equal(40);
+  });
+
+  it("should be able to execute opShr()", async function () {
+    const res = await shanghaiContract.opShr(2, 500);
+    // shift right
+    expect(res).to.equal(125);
+  });
+
+  it("should be able to execute opSar()", async function () {
+    const res = await shanghaiContract.opSar(1, 2);
+    // shift arithmetic right
+    expect(res).to.equal(1);
+  });
+
+  it("should be able to execute opExtCodeHash()", async function () {
+    const res = await shanghaiContract.opExtCodeHash(shanghaiContract.address);
+
+    // code hash
+    const prefix = res.toString().slice(0, 2);
+    expect(prefix).to.equal('0x');
+
+    const hash = res.toString().slice(2)
+    expect(hash.length).to.equal(64);
+
+    expect(res).not.to.equal(ethers.constants.HashZero)
+  });
+
+  it("should be able to execute opPush0()", async function () {
+    const res = await shanghaiContract.opPush0();
+    // push0
+    expect(res).to.equal(95);
```

Comment:
Suggest: Either use _hex_ notation here and in `ShanghaiOpcodes.sol` or use decimal notation in both places. Don't mix.  Confusing.  Unless you leave a comment behind in one place or the other.